packages/go_router/lib/src/builder.dart

```diff
@@ -47,164 +47,285 @@ class RouteBuilder {
     VoidCallback pop,
-    Key navigatorKey,
     bool routerNeglect,
   ) {
-    List<Page<dynamic>>? pages;
-    Exception? error;
-    final String location = matches.location.toString();
-    final List<RouteMatch> matchesList = matches.matches;
     try {
-      // build the stack of pages
-      if (routerNeglect) {
-        Router.neglect(
+      return tryBuild(
+          context, matchList, pop, routerNeglect, configuration.navigatorKey);
+    } on RouteBuilderError catch (e) {
+      return buildErrorNavigator(
           context,
-          () => pages = getPages(context, matchesList).toList(),
-        );
-      } else {
-        pages = getPages(context, matchesList).toList();
-      }
+          e,
+          Uri.parse(matchList.location.toString()),
+          pop,
+          configuration.navigatorKey);
+    }
+  }
 
-      // note that we need to catch it this way to get all the info, e.g. the
-      // file/line info for an error in an inline function impl, e.g. an inline
-      // `redirect` impl
-      // ignore: avoid_catches_without_on_clauses
-    } catch (err, stack) {
-      assert(() {
-        log.severe('Exception during GoRouter navigation', err, stack);
-        return true;
-      }());
-
-      // if there's an error, show an error page
-      error = err is Exception ? err : Exception(err);
-      final Uri uri = Uri.parse(location);
-      pages = <Page<dynamic>>[
-        _errorPageBuilder(
-          context,
-          GoRouterState(
-            configuration,
-            location: location,
-            subloc: uri.path,
-            name: null,
-            queryParams: uri.queryParameters,
-            error: error,
-          ),
-        ),
+  /// Builds the top-level Navigator by invoking the build method on each
+  /// matching route.
+  ///
+  /// Throws a [RouteBuilderError].
+  @visibleForTesting
+  Widget tryBuild(
+    BuildContext context,
+    RouteMatchList matchList,
+    VoidCallback pop,
+    bool routerNeglect,
+    GlobalKey<NavigatorState> navigatorKey,
+  ) {
+    return _buildRecursive(context, matchList, 0, pop, routerNeglect,
+        navigatorKey, <String, String>{}).widget;
+  }
+
+  /// Returns the top-level pages instead of the root navigator. Used for
+  /// testing.
+  @visibleForTesting
+  List<Page<dynamic>> buildPages(
+      BuildContext context, RouteMatchList matchList) {
+    try {
+      return _buildRecursive(context, matchList, 0, () {}, false,
+          GlobalKey<NavigatorState>(), <String, String>{}).pages;
+    } on RouteBuilderError catch (e) {
+      return <Page<dynamic>>[
+        buildErrorPage(context, e, matchList.location),
       ];
     }
+  }
 
-    // we should've set pages to something by now
-    assert(pages != null);
+  _RecursiveBuildResult _buildRecursive(
+    BuildContext context,
+    RouteMatchList matchList,
+    int startIndex,
+    VoidCallback pop,
+    bool routerNeglect, // TODO(johnpryan): Is this working? Needs a test...
```

Comment:
It works, but I was hoping a more flexible API instead of always ignores the browser history entry.

Reply:
👍 I'll remove this TODO